Change lattice version and get_family_data

murilobalves opened the pull request:
This PR must be accompanied with https://github.com/lnls-fac/pyaccel/pull/117, https://github.com/lnls-fac/mathphys/pull/25 and https://github.com/lnls-fac/trackcpp/pull/58

- <s>Set constant `lattice_version` in `__init__` instead of defining in accelerator_data dictionary in accelerator module</s> Include lattice_version as a property in `trackpp` and `pyaccel` accelerator classes. Now the string contains the accelerator version and also some repository information as: last tag, last commit and also if the repository was dirty when the accelerator was created. 
- <s>Delete unused accelerator_data dictionary from accelerator module (except for LI since it contains some parameters as emittance and energy spread, it could be deleted as well)</s> Accelerator_data dict was recovered
- Add if condition in `get_family_data()` for SI lattice so it works even for old lattices without definition of new IDs and corresponding correctors.

fernandohds564 commented on pymodels/SI_V25_04/families.py:
```diff
@@ -385,21 +385,28 @@ def get_idx(x):
     fams = ['IDC', ]
     for fam in fams:
-        idx.extend(data[fam])
+        if fam in data:
```
why do we need this if?

fernandohds564 reviewed (changes requested):
I think it's ok to set the variable `lattice_version` with the string of the lattice version, but I don't like the idea of removing the accelerator_data dictionary.

murilobalves commented:
> I think it's ok to set the variable `lattice_version` with the string of the lattice version, but I don't like the idea of removing the accelerator_data dictionary.

@fernandohds564 But the only information actually used in this accelerator data was the key `lattice_version` to define the homonymous module constant, that now would be included as a property in accelerator class. The other keys `global_coupling` and `pressure_profile` wasn't being used anywhere. I only kept `accelerator_data` on Linac since it includes emittance, energy spread and twiss values. I know that it is a non-uniformity wrt other accelerators, but LI is already different from the others anyways. Maybe we could use this accelerator_data to store useful information that we don't want to include elsewhere and actually use it along our codes. If you think that is the case, I can restore this dictionary definition.

fernandohds564 commented on pymodels/BO_V06_01/__init__.py:
```diff
@@ -17,4 +16,4 @@
 from .lattice import default_optics_mode, lattice_symmetry
 
-lattice_version = accelerator_data['lattice_version']
+lattice_version = 'BO_V06_01'
```
Why are we defining this constant in two places? I think we could define it only once and use its value, don't we?

Besides, wouldn't it be nice to have the version of the git repository plus the last commit appended to this string?

xresende replied:
I think it is better indeed to define the constant in only one place. But for me the correct place seems to be in the subpackage's __init__.py module. It seems natural to me something like

```Python3
from pymodels import SI_V25_04 as si

print(si.lattice_version)
```

and not 

```Python3

from pymodels import SI_V25_04 as si

print(si.lattice.lattice_version)
```

since everything with the subpackage `si`, and not only accelerator data, is particular to a particular model version (which defines the subpackage). 

I think murilo and I discussed leaving the constant also in `accelerator_data`in order not to break existing code. But we may fix them, as they show up.

murilobalves replied:
Thanks for your comments and suggestions! Based on them, I made some changes in the code and updated the PR description, please take a look.

xresende commented:
> Yes, I think this is the case. For example, it would be very useful to have real data in `pressure_profile` for lifetime calculations, or to have an `energy_acceptance` key for the same purposes.

@fernandohds564 , I agree this pressure profile is useful for beam lifetime calculation but ... It is not implemented this way yet. It is in `accelerator_data` because we copied and pasted from matlab scheme. It is so simple to implement in the future, when required for beam lifetime, that maybe it is worth not leaving it dangling in `accelerator_data`, raising questions. if pressure profile should be left at `accelerator_data` I think at least a comment is in order. Or maybe leave the line itself commented (as reminder that it can used in the future)

murilobalves commented:
I made several changes since the last review, mainly using the new functions implemented in mathphys (https://github.com/lnls-fac/mathphys/pull/25). Could you please review the changes again? Thanks!

The implementation of lattice_version definition includes information about the git repository (last tag, commit and if the repo is dirty) only if the pymodels package is installed with `make develop-install`. If the package is installed with `make install`, the package directory does not contain (yet) information about the version control of the corresponding repository, then only the module version is appended on lattice_version. 

@anacso17, @fernandohds564 and I discussed ways to include repo info at the moment the package was installed, in the case of `make install` is used. 
@anacso17 kindly agreed to be responsible of implementing and testing the changes on setup.py files of our packages to include this repo info appended on the package version. Once this task is achieved, I can return to pymodels and implement the definition of lattice_version including repo info also when `make install` is used.

fernandohds564 commented on pymodels/BO_V06_01/accelerator.py:
```diff
@@ -25,8 +28,23 @@ def create_accelerator(
     return accelerator
 
 
+lattice_version = 'BO_V06_01'
```
~~all this should come before you create the accelerator.~~
Actually I don't understand how this works, because you created a local variable named `lattice_version` inside the `create_accelerator` function and also another module variable with the same name and both of them have different values.

murilobalves replied:
I had forgotten to delete this local variable definition in `create_accelerator`, thanks!

fernandohds564 commented on pymodels/BO_V06_01/accelerator.py:
```diff
@@ -25,8 +28,23 @@ def create_accelerator(
 
+lattice_version = 'BO_V06_01'
+_path, _ver = _get_path('pymodels')
+if _is_git_repo(_path):
```
I think this creation of a standardized tag string should be a function of `mathphys` too.
Note how you had to repeat the code along all accelerators and how you had to import a lot of methods from `mathphys`.

murilobalves replied:
Sure, good idea. I implemented the method `get_package_string` in `mathphys`: https://github.com/lnls-fac/mathphys/pull/25/commits/cb1a8d36b9689df2336e0c559b70c80a108b08c0. Thanks!
